compiler/rustc_hir_typeck/src/fn_ctxt/adjust_fulfillment_errors.rs

```diff
@@ -166,11 +159,16 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             }
         }
 
-        match expr.map(|e| e.kind) {
+        match expr {
             Some(hir::ExprKind::MethodCall(segment, receiver, args, ..)) => {
-                if let Some(param) = direct_param
+                if let Some(param) = predicate_self_type_to_point_at
                     && self.point_at_generic_if_possible(error, def_id, param, segment)
                 {
+                    // HACK: This is not correct, since `predicate_self_type_to_point_at` might
```

Comment:
This shouldn't matter in practice, but I think it's important to note, since it can lead to weird diagnostics depending on how the `FunctionArgumentObligation` is used in the future.